lua/lualine/components/branch.lua

```diff
@@ -1,33 +1,79 @@
-local async = require('lualine.async')
-
 local git_branch
 
-local get_git_branch = async:new({
-  cmd = 'git branch --show-current',
-  on_stdout = function(_, data)
-    if data then
-      git_branch = data:gsub('\n', '')
+local known_git_dirs = {}
+local sep = package.config:sub(1,1)
+
+local function exists(path)
+  local f = io.open(path)
+  if not f then return false, false end
+  local _, _, errno = f:read()
+  f:close()
+  return true, errno == 21 -- is a directory
+end
+
+local function find_git_dir()
+  -- path seperator is not in the end add it
+  local file_path = vim.fn.expand('%:p:h')..sep
+  -- don't show branch for terminal buffers
+  if file_path:match('^term://.*$') then return nil end
+
+  local dir = file_path
+  -- do we need to iterate more than that?
+  local n = 30
+  while dir and n > 0 do
```

Comment:
Surely it would be better to have it search until $HOME universally on Linux, Mac and Windows. I'm not a fan of hardcoded variables and it will still be just as efficient while handling more edge cases (something like a huge java project may have a classic java directory hell)

Reply:
I was actually hoping to expose it as options once #60 was merged . 
Though I was using it to survive infinite loops during devlopment as geting stuck in inf loops in vim isn't nice as keyboard intrupt signals are captured by vim . It will just hang forever until killed.
Also I donn't like exposingnlocal functions as global function to call it from a event . I'll change it to a require statment one #60 is merged . For now I can only return one function.

I donn't think $HOME is a good idea . User might be using a different partition then where his/her home directory is . Also I donn't think windows has  a $HOME (might be wrong here).

Reply:
No way a user should have an option to change how this behaves. This will only produce issues down the line.

Please provide an example on how you plan on changing this to a require :)

 this still needs improvement, one reason why you could have been getting infinite loops was using dfs instead of bfs.
 you can check [rooter implementation for reference](https://github.com/airblade/vim-rooter/blob/master/plugin/rooter.vim) for inspiration

Reply:
```
vim.cmd[[autocmd BufEnter * :lua require'lualine.components.branch'.update_branch()]]
return {
  init = function(args) return branch(args) end,
  update_branch = watch_head}
```

I actually haven't gotten any inf loops ever since implementation was stable (about a week) . I was getting them when I was writting the function . I just left them just incase. It most likely can be safely removed.

Whats dfs and bfs?

Reply:
depth first search and breadth first search...

I hope we can remove this hardcoded value somehow without getting cpu heavy/infinite looping

Reply:
We are going up the directory tree not tranversing down. Does bfs have any value here at all?

Reply:
Okay wait, this whole `find_git_dir` function is traversing up in line `38` to find possible match and traversing down in line `45` if no match is found? Am I interpreting this correctly now?

Reply:
maybe your line numbers are off .
basically 
- check if dir/.git exists
- go up a dir
- repeat

Reply:
yeah yeah ok. 

If that's the case then everything works as expected, but if that's the case then I think the io calls in `exists` are useless since in path
```
a/b/c/d/e/file
```
The git repo will be in one of a,b,c,d,e and I think we know for sure that these are directories, however we still need to figure out a better ending point for this loop, better than `n=30`

Reply:
>I think the io calls in exists are useless since in path

I donn't know any other way to verify if a file exists or not If you know please suggest.

> however we still need to figure out a better ending point for this loop, better than n=30

I agree with that . but that was mostly to block inf loops . The thing is there might be some special char in filepath that may break the algorithm and cause inf loop . Since I donn't know all sort of filepath that are valid . I've left it there as a contingency

Reply:
I think it would be better to use `vim.fn.empty` and `vim.fn.isdirectory` since they are already implemented and are surely faster than io calls

Reply:
this whole function is actually reduced to `vim.fn.isdirectory(path)`

We can keep io calls in `get_git_head` since we actually need to read this file. 

This change will reduce io calls to a total of ONE :)